src/server/pkg/pool/pool.go

```diff
@@ -99,27 +104,32 @@ func (p *Pool) Do(ctx context.Context, f func(cc *grpc.ClientConn) error) error
 	if err := func() error {
 		p.connsLock.Lock()
 		defer p.connsLock.Unlock()
-		for addr, mapConn := range p.conns {
-			if mapConn.cc == nil {
-				cc, err := grpc.DialContext(ctx, addr, p.opts...)
-				if err != nil {
-					return err
-				}
-				mapConn.cc = cc
-				conn = mapConn
-				// We break because this conn has a count of 0 which we know
-				// we're not beating
-				break
-			} else {
-				if conn == nil || atomic.LoadInt64(&mapConn.count) < atomic.LoadInt64(&conn.count) {
+		for {
+			for addr, mapConn := range p.conns {
+				if mapConn.cc == nil {
+					cc, err := grpc.DialContext(ctx, addr, p.opts...)
+					if err != nil {
+						return err
+					}
+					mapConn.cc = cc
 					conn = mapConn
+					// We break because this conn has a count of 0 which we know
+					// we're not beating
+					break
+				} else {
+					mapCountCount := atomic.LoadInt64(&mapConn.count)
+					if mapCountCount < p.queueSize && (conn == nil || mapCountCount < atomic.LoadInt64(&conn.count)) {
+						conn = mapConn
+					}
 				}
 			}
+			if conn == nil {
+				p.connsCond.Wait()
```

Comment:
I could be missing something, but seems like the only way for this to unblock is for new workers to be discovered.  But shouldn't we unblock also when a worker's queue is no longer full?

Reply:
Ah, that's a very good point.

Reply:
Should be fixed, by line 137 below.